### DISCUSSION TITLE
Implement custom backend

### DISCUSSION THREAD

**Discussion by barsdeveloper** [SeaQL/sea-orm#2369] [Q&A] [answered]
Realistically, how hard would it be to implement a the backend for a custom database, one that is not mysql, postgres or sqlite? Is it supported by this ORM and if so where to start from?

**Comment by tyt2y3** [comment]
Realistically, this is very hard and a lot of work if the database is not already a variant of either MySQL or Postgres.
e.g. SQL Server etc has to be implemented from scratch almost.
The SQL syntax is different, so it starts from a custom backend in SeaQuery. And then a custom backend for SeaSchema, and finally some high-level abstractions in SeaORM, because each database has its own quirky features.

**Reply by barsdeveloper** [reply]
It's about DuckDB. Its SQL flavor is very similar to Postgres, it's rust driver it's very similar to SQLite (in fact it's forked from that one). Can you point me somewhere to start? Something like some documentation, some traits I need to implement etc

**Reply by barsdeveloper** [reply]
I've started to look into it bit by bit but I have an early obstacle that seems to have no solution. A connection object implements the trait [`ConnectionTrait`](https://docs.rs/sea-orm/latest/sea_orm/trait.ConnectionTrait.html) this one has a method that expects as result a [`ExecResult`](https://docs.rs/sea-orm/latest/sea_orm/struct.ExecResult.html). From this point there is no information in the documentation because what is inside is encapsulated, more specifically it has a field of type `ExecResultHolder` which is a pub(crate) enum that is not accessible from another crate.
Is there any plan to support DuckDB (please, it is very fast analytic database and very similar to sqlite that will get a lot of popularity) or to make it so that one can extend SeaORM?

**Reply by barsdeveloper** [reply]
@tyt2y3 Is there anything on the development roadmap about making SeaORM extendable to implement a custom database backend? I mean some sort of refactoring to solve problems as the one I talked about on the previous comment. Please note that sqlx seems to be able to be extendable to support new dbms without changing it: https://github.com/launchbadge/sqlx/issues/3186#issuecomment-2380380831
I've briefly looked into the documentation and it seems that their return types are expressed in terms of the Database associated type
https://docs.rs/sqlx/latest/sqlx/trait.Executor.html#method.execute
I think there must be a way to express all the return types in SeaORM in such a way to make it generic or reuse parts of sqlx so that a new backend could be implemented.

**Comment by tyt2y3** [comment, marked as answer]
> Is there any plan to support DuckDB 

DuckDB has always been on my radar! but we haven't got any plan for it yet.

> Is there anything on the development roadmap about making SeaORM extendable to implement a custom database backend

No, there isn't. SeaORM applies "enum based polymorphism", and enums are essentially closed ended and static. Custom backends implemented by third party have pitfalls in terms of feature coverage and update cycle. So I think the best way for the community is having us (you & I) to put everything together in the same repo and to be maintained holistically.

**Reply by barsdeveloper** [reply]
Thanks for the reply. Unfortunately as much as I like DuckDB probably more users would want to have MS SQL Server and maybe ClickHouse and so on. Supporting 3 backends is hard, supporting 10+ is close to impossible.
It makes sense to at least consider moving to traits and generics type of polymorphism.
